trie/database.go

```diff
@@ -30,64 +34,124 @@ type StateDB struct {
 	Hasher *Hasher
 }
 
-// WriteToDB writes the trie to the underlying database batch writer
-// Stores the merkle value of the node as the key and the encoded node as the value
-// This does not actually write to the db, just to the batch writer
-// Commit must be called afterwards to finish writing to the db
-func (t *Trie) WriteToDB() error {
-	t.db.Batch = t.db.Db.NewBatch()
-	return t.writeToDB(t.root)
+// Encode traverses the trie recursively, encodes each node, SCALE encodes the encoded node, and appends them all together
+func (t *Trie) Encode() ([]byte, error) {
+	return encode(t.root, []byte{})
 }
 
-// writeToDB recursively attempts to write each node in the trie to the db batch writer
-func (t *Trie) writeToDB(n node) error {
-	_, err := t.writeNodeToDB(n)
+func encode(n node, enc []byte) ([]byte, error) {
+	nenc, err := n.Encode()
 	if err != nil {
-		return err
+		return enc, err
+	}
+
+	scnenc, err := scale.Encode(nenc)
+	if err != nil {
+		return nil, err
 	}
 
+	enc = append(enc, scnenc...)
+
 	switch n := n.(type) {
 	case *branch:
 		for _, child := range n.children {
 			if child != nil {
-				err = t.writeToDB(child)
+				enc, err = encode(child, enc)
 				if err != nil {
-					return err
+					return enc, err
 				}
 			}
 		}
 	}
 
-	return nil
+	return enc, nil
 }
 
-// writeNodeToDB returns true if node is written to db batch writer, false otherwise
-// if node is clean, it will not attempt to be written to the db
-// otherwise if it's dirty, try to write it to db
-func (t *Trie) writeNodeToDB(n node) (bool, error) {
-	if !n.isDirty() {
-		return false, nil
+// Decode decodes a trie from the DB and sets the receiver to it
+// The encoded trie must have been encoded with EncodeForDB
+func (t *Trie) Decode(enc []byte) error {
```

Comment:
since the Trie struct also has a DB behind it, so this allows someone to make a trie with a DB and then get the encoding from the DB, and decode it back into the trie